frog/responsive-images.rkt

```diff
@@ -0,0 +1,117 @@
+         racket/file
+         racket/function
+         racket/list
+         racket/path
+         racket/port
+         racket/system
+         racket/string
+         "params.rkt"
+         "util.rkt"
+         "verbosity.rkt"
+         (only-in srfi/1 list-index))
+
+(provide make-responsive clean-resized-images magick-available?)
+
+(define magick-available? (find-executable-path "magick"))
+
+(module+ test
+  (require rackunit))
+
+(define (image-width path)
+  (define (magick-width path)
+    (string->number
+     (with-output-to-string
+       (λ ()
+         ; Fails when running from inside DrRacket on Mac OS X due to pathologic PATH
+         (system (string-append "identify -format %w \"" (path->string path) "\""))))))
+  (define get-width magick-width)
+  (get-width path))
+
+(module+ test
+  (when magick-available?
+    (check-eq? (image-width (string->path "../example/img/800px-image.gif")) 800)))
+
+(define/contract (resize-image in new-width out)
+  (path? number? path? . -> . boolean?)
+  ;; Depend on ImageMagick
+  (define (magick-resize in w out)
+    (system (format "magick \"~a\" -geometry ~a \"~a\"" (path->string in) w (path->string out))))
+  (define resize magick-resize)
+  (prn1 "Resizing image ~a to ~a pixels... " (path->string in) new-width)
+  (if (> (current-verbosity) 1)
+      (time (resize in new-width out))
+      (resize in new-width out)))
+
+(module+ test
+  (when magick-available?
+    (define out (string->path "/tmp/resized.gif"))
+    (test-eq? "resize" (begin
+                         (resize-image (string->path "../example/img/480px-image.gif") 10 out)
+                         (image-width out))
+              10)
+    (delete-file out))) ; catch exn?
+
+(define/contract (get-images image-path)
+  (path? . -> . (values pair? (listof pair?)))
+  (define output-path (string->path (current-image-output-dir)))
+  (unless (directory-exists? output-path)
+    (make-directory output-path))
+  (let* ([orig-size (image-width image-path)]
+         [sizes (filter ((curry >=) orig-size) (current-image-sizes))])
+    (values (cons image-path orig-size)
+            (for/list ([width sizes])
+              (let* ([filename (file-name-from-path image-path)]
+                     [extension (get-extension filename)]
+                     [out-path
+                      (build-path
+                       output-path
+                       (replace-extension
+                        filename
+                        (string-append "-" (number->string width) "." extension)))])
+                (unless (and (file-exists? out-path)
+                             (< (file-or-directory-modify-seconds image-path)
+                                (file-or-directory-modify-seconds out-path)))
+                  (resize-image image-path width out-path))
+                (cons out-path width))))))
+
+(define default-image-idx
+  (list-index ((curry =) (current-image-default-size)) (current-image-sizes)))
```

Comment:
Rather than `require` Scheme `srfi/1` just to get `list-index`, I think it might be "more Rackety" to do something like:

```rkt
  (for/or ([v  (current-image-sizes)]
           [ix (in-naturals)])
    (and (= v (current-image-default-size)) ix))
```